chrome/content/zotero/xpcom/file.js

```diff
@@ -447,6 +447,46 @@ Zotero.File = new function(){
 	};
 	
 	
+	/**
+	 * Write a binary string to the given path
+	 *
+	 * This snippet comes from the example (Writing a Binary File) here:
+	 * https://developer.mozilla.org/en-US/docs/Archive/Add-ons/Code_snippets/File_I_O
+	 *
+	 * @param {String} path - Path to file
+	 * @param {String} data - Binary string to write to file
+	 * @return {Promise} - A promise that is resolved when the file has been written
+	 */
+	this.putBinaryContentsAsync = async function (path, data) {
+		await new Zotero.Promise(function (resolve, reject) {
+			try {
+				var aFile = Components.classes["@mozilla.org/file/local;1"]
+				.createInstance(Components.interfaces.nsIFile);
+				
+				aFile.initWithPath(path);
+				aFile.createUnique(Components.interfaces.nsIFile.NORMAL_FILE_TYPE, 0o755);
+				
+				var stream = Components.classes["@mozilla.org/network/safe-file-output-stream;1"]
+				.createInstance(Components.interfaces.nsIFileOutputStream);
+				stream.init(aFile, 0x04 | 0x08 | 0x20, 0o755, 0); // readwrite, create, truncate
+				
+				stream.write(data, data.length);
+				if (stream instanceof Components.interfaces.nsISafeOutputStream) {
+					stream.finish();
+				}
+				else {
+					stream.close();
+				}
+				resolve();
+			}
+			catch (e)
+			{
+				reject(e);
+			}
+		});
+	};
```

Comment:
Nothing in here is async, so if we were to keep this, it would be `putBinaryContents()`, and the `async`/`await`/`try`/`catch` would need to be removed. But `putContentsAsync()` should already handle this. It can take an input stream, ArrayBuffer, or Blob, and there's an existing test, "should save a Blob", for writing a binary blob.

Reply:
I spent some time on the MDN site looking for a stream that I could use to write a binary string using putContentsAsync, but had no luck. I tried this:
```
Components.utils.importGlobalProperties(["Blob"]);
let blob = new Blob([resource], { type: contentType });
await Zotero.File.putContentsAsync(
	filePath,
	blob
);
```
but the file was corrupted. This will depend on the data coming out of the processEndpoint so maybe I will wait until we get that settled first unless you see something obvious I'm missing?

Reply:
`putContentsAsync()` will happily take a Blob with valid data. E.g., this will create an exact copy:

```javascript
var data = await File.createFromFileName('/Users/dan/Desktop/z.png');
await Zotero.File.putContentsAsync('/Users/dan/Desktop/z2.png', data);
```

If `resource` is a binary string in `new Blob([resource]…)`, that won't work — you would have to convert that to a typed array, I think, to keep it from being interpreted as UTF-8 — but don't you have a Blob from the Connector now anyway?

Reply:
Ah right. `resource` is just a binary string from the multipart data so I can use charCodeAt and then convert it to a Blob. Thanks.